src/fsharp/DependencyManager.Integration.fs

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+/// Helper members to integrate DependencyManagers into F# codebase
+module internal FSharp.Compiler.DependencyManagerIntegration
+
+open System
+open System.Reflection
+open System.IO
+open FSharp.Compiler.ErrorLogger
+open Internal.Utilities.FSharpEnvironment
+open FSharp.Compiler.DotNetFrameworkDependencies
+
+// Contract strings
+let dependencyManagerPattern = "*DependencyManager*.dll"
+let dependencyManagerAttributeName= "DependencyManagerAttribute"
+let resolveDependenciesMethodName = "ResolveDependencies"
+let namePropertyName = "Name"
+let keyPropertyName = "Key"
+
+module ReflectionHelper =
+    let assemblyHasAttribute (theAssembly: Assembly) attributeName =
+        try
+            CustomAttributeExtensions.GetCustomAttributes(theAssembly)
+            |> Seq.exists (fun a -> a.GetType().Name = attributeName)
+        with | _ -> false
+
+    let getAttributeNamed (theType: Type) attributeName =
+        try
+            theType.GetTypeInfo().GetCustomAttributes false
+            |> Seq.tryFind (fun a -> a.GetType().Name = attributeName)
+        with | _ -> None
+
+    let getInstanceProperty<'treturn> (theType: Type) propertyName =
+        try
+            let property = theType.GetProperty(propertyName, typeof<'treturn>)
+            if isNull property then
+                None
+            elif not (property.GetGetMethod().IsStatic)
+                 && property.GetIndexParameters() = Array.empty
+            then
+                Some property
+            else
+                None
+        with | _ -> None
+
+    let getInstanceMethod<'treturn> (theType: Type) (parameterTypes: Type array) methodName =
+        try
+            let theMethod = theType.GetMethod(methodName, parameterTypes)
+            if isNull theMethod then
+                None
+            else
+                Some theMethod
+        with | _ -> None
+
+    let stripTieWrapper (e:Exception) =
+        match e with
+        | :? TargetInvocationException as e->
+            e.InnerException
+        | _ -> e
+
+(* Shape of Dependency Manager contract, resolved using reflection *)
+type internal IDependencyManagerProvider =
+    abstract Name : string
+    abstract Key: string
+    abstract ResolveDependencies : scriptDir: string * mainScriptName: string * scriptName: string * packageManagerTextLines: string seq * tfm: string -> bool * string list * string list
```

Comment:
Consider replacing the `string list` with arrays or enumerables. this makes it possible to implement the provider without dependency on FSharp.Core.

This was a concern when I made the initial implementation of the provider interface so the same provider could be loaded from scriptcs (@glennblock), ideally the same interface could just work for C# interactive as well (given they implement similar hooks at some point).

The signature should be documented, it is not obvious that return type is:
* success: bool
* load scripts: string []
* include dirs: string []